core/src/main/scala/cats/UnorderedFoldable.scala

```diff
@@ -0,0 +1,66 @@
+package cats
+
+import cats.kernel.CommutativeMonoid
+import simulacrum.typeclass
+
+import scala.collection.mutable
+
+/**
+  * `UnorderedFoldable` is like a `Foldable` for unordered containers.
+  */
+@typeclass trait UnorderedFoldable[F[_]] {
+
+  /**
+    * Left associative fold on 'F' using the function 'f'.
+    */
+  def foldLeft[A, B](fa: F[A], b: B)(f: (B, A) => B): B
+
+  def unorderedFoldMap[A, B: CommutativeMonoid](fa: F[A])(f: A => B): B =
+    foldLeft(fa, Monoid[B].empty)((b, a) => Monoid[B].combine(b, f(a)))
+
+  def unorderedFold[A: CommutativeMonoid](fa: F[A]): A =
+    unorderedFoldMap(fa)(identity)
+
+  def toSet[A](fa: F[A]): Set[A] =
```

Comment:
This can be implemented with foldMap

Reply:
I wonder if actually we should take an Order here. This is using universal hashing and i hate to bake that into a new typeclass.

Reply:
If we take an Order we could make it a `SortedSet`, not sure if that really makes sense. I agree with your sentiment, but not sure how to deal with it exactly. Maybe take a `Hash[A]`? I dunno.

Reply:
I agree that having `UnorderedFoldable` taking a `Order` doesn't make much sense. 
The whole point of this PR is to support a principled functionalities for `Set` and `Map`, which are based off universal hashing I am not sure if we can get away from it. 
If we really want `UndorderedFoldable` as a principled abstraction,  we should probably only support the `Set` and `Map` in dogs.

Reply:
we could make this `toList` with the only law being about `contains` and `Eq[A]`. It would be `O(N^2)` to check the law, but that is probably fine.

I just hate sneaking universal hashing or universal equality in. 

I don't see the use or `Order[A]` as not making sense, since it is the *collection* that is unordered here, not the type `A`. So, I don't want to get those two confused.

Reply:
+1 on this solution.

Reply:
I'm not sure `toList` really makes sense, given the fact that we only want to fold using a `CommutativeMonoid` which `List` clearly is not. Maybe we just get rid of this method all together? I really wish we had a `HashMap`.

Reply:
I agree. At the very least this is not stable enough to be included in 1.0.0